query_optimizer/rules/BottomUpRule.hpp

```diff
@@ -80,6 +81,8 @@ class BottomUpRule : public Rule<TreeType> {
    */
   virtual TreeNodePtr applyToNode(const TreeNodePtr &node) = 0;
 
+  virtual void init(const TreeNodePtr &node) = 0;
```

Comment:
Let's provide a default no-op implementation for `init()` here (instead of pure virtual) so we don't have to implement it for every subclass.